src/Deinterleave.cpp

```diff
@@ -578,173 +578,247 @@ class Interleaver : public IRMutator {
-
-        // Not enough stores collected.
-        if (stores.size() != (size_t)expected_stores) return Stmt();
-
-        Type t = store->value.type();
-        Expr base;
-        std::vector<Expr> args(stores.size());
-        std::vector<Expr> predicates(stores.size());
-
-        int min_offset = 0;
-        std::vector<int> offsets(stores.size());
+            Type t = c.type;
+            Expr base;
+            std::vector<Expr> args(stores.size());
+            std::vector<Expr> predicates(stores.size());
+
+            int min_offset = 0;
+            std::vector<int> offsets(stores.size());
+
+            std::string load_name;
+            Buffer<> load_image;
+            Parameter load_param;
+            bool failed = false;
+            for (size_t i = 0; i < stores.size(); ++i) {
+                const Ramp *ri = stores[i].as<Store>()->index.as<Ramp>();
+                internal_assert(ri);
+
+                // Mismatched store vector laness.
+                if (ri->lanes != c.type.lanes()) {
+                    failed = true;
+                    break;
+                }
 
-        std::string load_name;
-        Buffer<> load_image;
-        Parameter load_param;
-        for (size_t i = 0; i < stores.size(); ++i) {
-            const Ramp *ri = stores[i].as<Store>()->index.as<Ramp>();
-            internal_assert(ri);
+                Expr diff = simplify(ri->base - c.base);
+                const int64_t *offs = as_const_int(diff);
 
-            // Mismatched store vector laness.
-            if (ri->lanes != lanes) return Stmt();
+                // Difference between bases is not constant.
+                if (!offs) {
+                    failed = true;
+                    break;
+                }
 
-            Expr diff = simplify(ri->base - r0->base);
-            const int64_t *offs = as_const_int(diff);
+                offsets[i] = *offs;
+                if (*offs < min_offset) {
+                    min_offset = *offs;
+                }
 
-            // Difference between bases is not constant.
-            if (!offs) return Stmt();
+                if (c.stride == 1) {
+                    // Difference between bases is not a multiple of the lanes.
+                    if (*offs % c.type.lanes() != 0) {
+                        failed = true;
+                        break;
+                    }
+
+                    // This case only triggers if we have an immediate
+                    // load of the correct stride on the RHS.  TODO:
+                    // Could we consider mutating the RHS so that we
+                    // can handle more complex Expr's than just loads?
+                    const Load *load = stores[i].as<Store>()->value.as<Load>();
+                    if (!load) {
+                        failed = true;
+                        break;
+                    }
+
+                    // TODO(psuriana): Predicated load is not currently handled.
```

Comment:
Is this really the right assignee for this TODO?

Reply:
old code moved around. I'll at least delete the assignee